src/core/search/ast_expr.h

```diff
@@ -4,21 +4,106 @@
 
 #pragma once
 
+#include <algorithm>
+#include <iostream>
+#include <memory>
 #include <ostream>
+#include <vector>
+
+#include "base/logging.h"
 
 namespace dfly {
 
 namespace search {
 
-class AstExpr {};
+// Describes a single node of the filter AST tree.
+class AstNode {
+ public:
+  virtual ~AstNode() = default;
+
+  // Check if this input is matched by the node.
+  virtual bool Check(std::string_view input) const = 0;
+
+  // Debug print node.
+  virtual std::string Debug() const = 0;
+};
+
+using AstNodePtr = std::unique_ptr<AstNode>;
+
+// Do we have to pollute all our parser file with moves?
+struct AstExpr {
+  AstExpr() : ptr{} {
+  }
+  AstExpr(AstNodePtr ptr) : ptr{move(ptr)} {
+  }
+  AstExpr(const AstExpr& other) {
+    ptr = move(other.ptr);
+  }
+  AstExpr& operator=(const AstExpr& other) {
+    ptr = move(other.ptr);
+    return *this;
+  }
+  operator AstNodePtr() {
+    return move(ptr);
+  }
+  const AstNode* operator->() const {
+    return ptr.get();
+  }
+  mutable AstNodePtr ptr;
+};
+
+template <typename T, typename... Ts> AstExpr MakeExpr(Ts... ts) {
+  AstNode* node = dynamic_cast<AstNode*>(new T(std::forward<Ts>(ts)...));
+  CHECK(node);
+  return std::unique_ptr<AstNode>{node};
+}
+
+// AST term node, matches only if input contains term.
+class AstTermNode : public AstNode {
+ public:
+  AstTermNode(std::string term) : term_{move(term)} {
+  }
+  virtual bool Check(std::string_view input) const;
+  virtual std::string Debug() const;
+
+ private:
+  std::string term_;
+};
+
+// Ast negation node, matches only if its sub node didn't match.
+class AstNegateNode : public AstNode {
+ public:
+  AstNegateNode(AstNodePtr node) : node_{move(node)} {
+  }
+  virtual bool Check(std::string_view needle) const;
+  virtual std::string Debug() const;
+
+ private:
+  AstNodePtr node_;
+};
+
+// Ast logical operation node, matches only if sub nodes match
+// in respect to logical operation (and/or).
+class AstLogicalNode : public AstNode {
+ public:
+  AstLogicalNode(AstNodePtr l, AstNodePtr r, bool disjunction)
+      : l_{move(l)}, r_{move(r)}, disjunction_{disjunction} {
+  }
+  virtual bool Check(std::string_view needle) const;
+  virtual std::string Debug() const;
+
+ private:
+  AstNodePtr l_, r_;
+  bool disjunction_;
+};
 
 }  // namespace search
 }  // namespace dfly
 
 namespace std {
```

Comment:
Not part of your PR, but Google style guide forbids declaring stuff in the `std` namespace:

> Do not declare anything in namespace std, including forward declarations of standard library classes. Declaring entities in namespace std is undefined behavior, i.e., not portable


Source: https://google.github.io/styleguide/cppguide.html#Namespaces

I think that the best practice is to have a `friend ostream& operator<<` inside the class

Reply:
@chakaz for this specific case, we historically used such declarations but if there is a cleaner solution - lets adopt it.

Reply:
yep. one additional small benefit with hidden friends is keeping the overload set for `operator<<` smaller, so the compiler is slightly faster and the compilation errors for unprintable types are shorter :)

Reply:
I think that it's cleaner and more encapsulated to use the `friend` thingy:

```c++
#include <iostream>
using namespace std;

struct MyClass {
  friend std::ostream& operator<<(std::ostream& o, const MyClass& t) {
    o << "MyClass{}";
    return o;
  }
};

int main() {
  MyClass c;
  cout << "Hello, World!" << c << endl;
}
```